File: src/main/java/com/github/samarium150/structurescompass/gui/StructuresCompassHUD.java
package com.github.samarium150.structurescompass.gui;

import com.github.samarium150.structurescompass.config.StructuresCompassConfig;
import com.github.samarium150.structurescompass.init.ItemRegistry;
import com.github.samarium150.structurescompass.item.StructuresCompassItem;
import com.github.samarium150.structurescompass.util.RenderUtils;
import com.github.samarium150.structurescompass.util.StructureUtils;
import com.mojang.blaze3d.vertex.PoseStack;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.GuiComponent;
import net.minecraft.client.gui.screens.ChatScreen;
import net.minecraft.client.resources.language.I18n;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.ItemStack;
import net.minecraft.core.BlockPos;
import net.minecraft.world.phys.Vec3;

/**
 * The HUD of the compass
 */
public final class StructuresCompassHUD extends GuiComponent {
    
    private static final Minecraft minecraft = Minecraft.getInstance();
    private final PoseStack matrixStack;
    
    /**
     * Initializer of the HUD
     * @param matrixStack MatrixStack from RenderGameOverlayEvent
     * @see net.minecraftforge.client.event.RenderGameOverlayEvent
     */
    public StructuresCompassHUD(PoseStack matrixStack) {
        this.matrixStack = matrixStack;
    }
    
    /**
     * Render the HUD
     * <p>
     * This method is adapted from
     * <a href="https://github.com/MattCzyr/NaturesCompass" target="_blank">NaturesCompass</a>
     * which is under
     * <a href="https://creativecommons.org/licenses/by-nc-sa/4.0" target="_blank">
     * Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License
     * </a>.
     */
    public void render() {
        
        if (StructuresCompassConfig.HUD_Level.get() != 0 &&
                (minecraft.screen == null || (StructuresCompassConfig.displayWithChatOpen.get()
                                                            && minecraft.screen instanceof ChatScreen))) {
            final Player player = minecraft.player;
            final ItemStack stack;
            assert player != null;
            if (!player.getMainHandItem().isEmpty()
                    && player.getMainHandItem().getItem() == ItemRegistry.STRUCTURES_COMPASS.get())
                stack = player.getMainHandItem();
            else if (!player.getOffhandItem().isEmpty()
                         && player.getOffhandItem().getItem() == ItemRegistry.STRUCTURES_COMPASS.get())
                stack = player.getOffhandItem();
            else return;
            if (stack.getItem() instanceof StructuresCompassItem) {
                String feature = StructuresCompassItem.getStructureName(stack);
                String dim = StructuresCompassItem.getDimension(stack);
                BlockPos pos = StructuresCompassItem.getPos(stack);
                int relLineOffset = 0;
                if (feature == null) return;
                if (pos == null || dim == null) {
                    RenderUtils.drawConfiguredStringOnHUD(
                        matrixStack, I18n.get("string.structurescompass.hud_not_found"),
                        5, 5, 0xAA2116, relLineOffset
                    );
                } else {
                    RenderUtils.drawConfiguredStringOnHUD(
                        matrixStack, I18n.get("string.structurescompass.hud_structure"),
                        5, 5, 0xFFFFFF, relLineOffset);
                    RenderUtils.drawConfiguredStringOnHUD(
                        matrixStack, StructureUtils.getLocalizedStructureName(feature),
                        5, 5, 0xAAAAAA, ++relLineOffset);
                    relLineOffset++;
            
                    RenderUtils.drawConfiguredStringOnHUD(
                        matrixStack, I18n.get("string.structurescompass.hud_dim"),
                        5, 5, 0xFFFFFF, ++relLineOffset
                    );
                    RenderUtils.drawConfiguredStringOnHUD(
                        matrixStack, StructureUtils.getLocalizedDimensionName(dim),
                        5, 5, 0xAAAAAA, ++relLineOffset
                    );
                    relLineOffset++;

                    if (StructuresCompassConfig.HUD_Level.get() == 3) {
                        String temp = pos.getY() == 0 ? ", X, " : ", " + pos.getY() + ", ";
                        temp = I18n.get("string.structurescompass.hud_pos") + " [" + pos.getX() + temp + pos.getZ() + "]";
                        RenderUtils.drawConfiguredStringOnHUD(
                            matrixStack, temp,
                            5, 5, 0x4AFF4A, ++relLineOffset
                        );
                    }
            
                    if (player.getCommandSenderWorld().dimension().location().toString().equals(dim)) {
                
                        Vec3 dis = StructureUtils.getDistance(pos, player);
                        double disX = dis.x();
                        double disY = dis.y();
                        double disZ = dis.z();
                        double distance = (double) Math.round(Math.sqrt(disX * disX + disY * disY + disZ * disZ) * 100) / 100;
                
                        if (StructuresCompassConfig.HUD_Level.get() == 3 && disX > StructuresCompassConfig.closeEnough.get())
                            RenderUtils.drawConfiguredStringOnHUD(
                                matrixStack, I18n.get("string.structurescompass.hud_east") + disX,
                                5, 5, 0xFFFFFF, ++relLineOffset
                            );
                        else if (StructuresCompassConfig.HUD_Level.get() == 3 && disX < -StructuresCompassConfig.closeEnough.get())
                            RenderUtils.drawConfiguredStringOnHUD(
                                matrixStack, I18n.get("string.structurescompass.hud_west") + -disX,
                                5, 5, 0xFFFFFF, ++relLineOffset
                            );
                
                        if (StructuresCompassConfig.HUD_Level.get() == 3 && disZ > StructuresCompassConfig.closeEnough.get())
                            RenderUtils.drawConfiguredStringOnHUD(
                                matrixStack, I18n.get("string.structurescompass.hud_south") + disZ,
                                5, 5, 0xFFFFFF, ++relLineOffset
                            );
                        else if (StructuresCompassConfig.HUD_Level.get() == 3 && disZ < -StructuresCompassConfig.closeEnough.get())
                            RenderUtils.drawConfiguredStringOnHUD(
                                matrixStack, I18n.get("string.structurescompass.hud_north") + -disZ,
                                5, 5, 0xFFFFFF, ++relLineOffset
                            );
                
                        if (StructuresCompassConfig.HUD_Level.get() == 3 && disY > StructuresCompassConfig.closeEnough.get())
                            RenderUtils.drawConfiguredStringOnHUD(
                                matrixStack, I18n.get("string.structurescompass.hud_up") + disY,
                                5, 5, 0xFFFFFF, ++relLineOffset
                            );
                        else if (StructuresCompassConfig.HUD_Level.get() == 3 && disY < -StructuresCompassConfig.closeEnough.get())
                            RenderUtils.drawConfiguredStringOnHUD(
                                matrixStack, I18n.get("string.structurescompass.hud_down") + -disY,
                                5, 5, 0xFFFFFF, ++relLineOffset
                            );
                
                        if ((StructuresCompassConfig.HUD_Level.get() == 3 && relLineOffset > 6) ||
                            (StructuresCompassConfig.HUD_Level.get() >= 2 && distance > StructuresCompassConfig.closeEnough.get())) {
                            relLineOffset++;
                            RenderUtils.drawConfiguredStringOnHUD(
                                matrixStack, I18n.get("string.structurescompass.hud_distance") + distance,
                                5, 5, 0xFFC20E, ++relLineOffset
                            );
                        }
                    } else {
                        relLineOffset++;
                        RenderUtils.drawConfiguredStringOnHUD(
                            matrixStack, I18n.get("string.structurescompass.hud_wrong_dim"),
                            5, 5, 0xAA2116, ++relLineOffset
                        );
                    }
                }
            }
        }
    }
}

File: src/main/kotlin/io/github/samarium150/minecraft/mod/structures_compass/client/gui/hud/StructuresCompassHud.kt
/*
 * Copyright (c) 2022 Samarium
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/gpl-3.0.html>.
 */
package io.github.samarium150.minecraft.mod.structures_compass.client.gui.hud

import io.github.samarium150.minecraft.mod.structures_compass.client.util.drawConfiguredStringOnHud
import io.github.samarium150.minecraft.mod.structures_compass.client.util.getLocalizedDimensionName
import io.github.samarium150.minecraft.mod.structures_compass.client.util.getLocalizedStructureName
import io.github.samarium150.minecraft.mod.structures_compass.config.ClientConfig
import io.github.samarium150.minecraft.mod.structures_compass.config.StructuresCompassConfig
import io.github.samarium150.minecraft.mod.structures_compass.init.ItemRegistry
import io.github.samarium150.minecraft.mod.structures_compass.util.*
import net.fabricmc.api.EnvType
import net.fabricmc.api.Environment
import net.minecraft.client.font.TextRenderer
import net.minecraft.client.gui.screen.ChatScreen
import net.minecraft.client.resource.language.I18n
import net.minecraft.client.util.math.MatrixStack

@Environment(EnvType.CLIENT)
object StructuresCompassHud {

    private val textRenderer: TextRenderer
        get() = minecraftClient.textRenderer

    private val config: ClientConfig
        get() = StructuresCompassConfig.configData.client

    fun render(matrixStack: MatrixStack) {
        if (minecraftClient.currentScreen == null ||
            (config.displayWithChatOpen && minecraftClient.currentScreen is ChatScreen)) {
            val player = minecraftClient.player
            if (player != null) {
                val itemStack = if (player.mainHandStack.item == ItemRegistry.STRUCTURES_COMPASS) {
                    player.mainHandStack
                } else if (player.offHandStack.item == ItemRegistry.STRUCTURES_COMPASS) {
                    player.offHandStack
                } else return
                val structureId = itemStack.getStructure()
                val dimension = itemStack.getDimension()
                val position = itemStack.getPos()
                var relLineOffset = 0
                if (structureId == null) return
                if (position == null || dimension == null) {
                    textRenderer.drawConfiguredStringOnHud(
                        matrixStack,
                        I18n.translate("${prefix}hud_not_found"),
                        5, 5, 0xAA2116, relLineOffset
                    )
                } else {
                    textRenderer.drawConfiguredStringOnHud(
                        matrixStack,
                        I18n.translate("${prefix}hud_structure"),
                        5, 5, 0xFFFFFF, relLineOffset
                    )
                    textRenderer.drawConfiguredStringOnHud(
                        matrixStack,
                        structureId.getLocalizedStructureName(),
                        5, 5, 0xAAAAAA, ++relLineOffset
                    )
                    relLineOffset++

                    textRenderer.drawConfiguredStringOnHud(
                        matrixStack,
                        I18n.translate("${prefix}hud_dim"),
                        5, 5, 0xFFFFFF, ++relLineOffset
                    )
                    textRenderer.drawConfiguredStringOnHud(
                        matrixStack,
                        dimension.getLocalizedDimensionName(),
                        5, 5, 0xAAAAAA, ++relLineOffset
                    )
                    relLineOffset++

                    if (config.HudInfoLevel == 3)
                        textRenderer.drawConfiguredStringOnHud(
                            matrixStack, I18n.translate("${prefix}hud_pos") +
                                " [${position.x}, ${if (position.y == 0) "X" else "${position.y}"}, ${position.z}]",
                            5, 5, 0x4AFF4A, ++relLineOffset
                        )

                    if (player.world.registryKey.value == dimension) {
                        val distanceVector = position.getDistanceVector(player)
                        val x = distanceVector.x
                        val y = distanceVector.y
                        val z = distanceVector.z
                        val distance = distanceVector.getLength()
                        if (config.HudInfoLevel == 3) {
                            if (x > 0.3)
                                textRenderer.drawConfiguredStringOnHud(
                                    matrixStack, "${I18n.translate("${prefix}hud_east")} $x",
                                    5, 5, 0xFFFFFF, ++relLineOffset
                                )
                            else if (x < -0.3)
                                textRenderer.drawConfiguredStringOnHud(
                                    matrixStack, "${I18n.translate("${prefix}hud_west")} ${-x}",
                                    5, 5, 0xFFFFFF, ++relLineOffset
                                )
                            if (z > 0.3)
                                textRenderer.drawConfiguredStringOnHud(
                                    matrixStack, "${I18n.translate("${prefix}hud_south")} $z",
                                    5, 5, 0xFFFFFF, ++relLineOffset
                                )
                            else if (z < -0.3)
                                textRenderer.drawConfiguredStringOnHud(
                                    matrixStack, "${I18n.translate("${prefix}hud_north")} ${-z}",
                                    5, 5, 0xFFFFFF, ++relLineOffset
                                )
                            if (y > 0.3)
                                textRenderer.drawConfiguredStringOnHud(
                                    matrixStack, "${I18n.translate("${prefix}hud_up")} $y",
                                    5, 5, 0xFFFFFF, ++relLineOffset
                                )
                            else if (y < -0.3)
                                textRenderer.drawConfiguredStringOnHud(
                                    matrixStack, "${I18n.translate("${prefix}hud_down")} ${-y}",
                                    5, 5, 0xFFFFFF, ++relLineOffset
                                )
                        }
                        if ((config.HudInfoLevel == 3 && relLineOffset > 6) ||
                            (config.HudInfoLevel >= 3 && distance > config.closedEnough)) {
                            relLineOffset++
                            textRenderer.drawConfiguredStringOnHud(
                                matrixStack, "${I18n.translate("${prefix}hud_distance")} %.3f".format(distance),
                                5, 5, 0xFFC20E, ++relLineOffset
                            )
                        }
                    } else {
                        relLineOffset++
                        textRenderer.drawConfiguredStringOnHud(
                            matrixStack, I18n.translate("${prefix}hud_wrong_dim"),
                            5, 5, 0xAA2116, ++relLineOffset
                        )
                    }
                }
            }
        }
    }
}
